homeassistant/components/fireservicerota/sensor.py

```diff
@@ -0,0 +1,131 @@
+        self._entry_id = entry.entry_id
+        self._unique_id = entry.unique_id
+
+        self._state = None
+        self._state_attributes = {}
+
+    @property
+    def name(self) -> str:
+        """Return the name of the sensor."""
+        return "Incidents"
+
+    @property
+    def icon(self) -> str:
+        """Return the icon to use in the frontend."""
+        return "mdi:fire-truck"
+
+    @property
+    def state(self) -> str:
+        """Return the state of the sensor."""
+        return self._state
+
+    @property
+    def unique_id(self) -> str:
+        """Return the unique ID for this sensor."""
+        return f"{self._unique_id}_Incidents"
+
+    @property
+    def device_state_attributes(self) -> object:
+        """Return available attributes for sensor."""
+        attr = {}
+        data = self._state_attributes
+
+        if data:
+            for value in (
+                "trigger",
+                "created_at",
+                "message_to_speech_url",
+                "prio",
+                "type",
+                "responder_mode",
+                "can_respond_until",
+            ):
+                if data.get(value):
+                    attr[value] = data[value]
+
+            if "address" in data:
+                for address_value in (
+                    "latitude",
+                    "longitude",
+                    "address_type",
+                    "formatted_address",
+                ):
+                    if address_value in data["address"]:
+                        attr[address_value] = data["address"][address_value]
+
+            attr[ATTR_ATTRIBUTION] = ATTRIBUTION
+            return attr
+
+    async def async_added_to_hass(self) -> None:
+        """Handle entity which will be added."""
+        await super().async_added_to_hass()
+
+        state = await self.async_get_last_state()
+        if state:
+            self._state = state.state
+            self._state_attributes = state.attributes
+
+        self.async_on_remove(self._coordinator.stop_listener)
+
+        self.async_on_remove(
+            async_dispatcher_connect(
+                self.hass,
+                f"{FIRESERVICEROTA_DOMAIN}_{self._entry_id}_update",
+                self.async_on_demand_update,
+            )
+        )
+
+    @property
+    def should_poll(self) -> bool:
+        """No polling needed."""
+        return False
+
+    async def async_on_demand_update(self) -> None:
+        """Update state on demand."""
+        self.async_schedule_update_ha_state(True)
+
+    async def async_update(self) -> None:
+        """Update using FireServiceRota data."""
+        if not self.enabled:
+            return
+
+        try:
+            self._state = self._coordinator.incident_data["body"]
+            self._state_attributes = self._coordinator.incident_data
+        except (KeyError, TypeError):
+            pass
+
+        _LOGGER.debug("Entity 'Incidents' state set to: %s", self._state)
```

Comment:
I don't think you need the async_on_demand_update, you could instead on line 107 you could set it to call async_update and make the update method call self.async_write_ha_state() at the end

Reply:
Good suggestion, it was copy and pasted code from another integration.
I now call self.async_write_ha_state() inside the try-except code though, if i put it at the end above the logger line I get below error at restart (when state is restored) does this mean there is a bug somewhere? Or expected behavior?
```
2020-10-16 08:27:48 DEBUG (MainThread) [homeassistant.components.fireservicerota] Updating availability data
2020-10-16 08:27:48 ERROR (MainThread) [homeassistant.components.sensor] fireservicerota: Error on device update!
Traceback (most recent call last):
  File "/workspaces/core/homeassistant/helpers/entity_platform.py", line 321, in _async_add_entity
    await entity.async_device_update(warning=False)
  File "/workspaces/core/homeassistant/helpers/entity.py", line 460, in async_device_update
    await self.async_update()  # type: ignore
  File "/workspaces/core/homeassistant/components/fireservicerota/sensor.py", line 127, in async_update
    self.async_write_ha_state()
  File "/workspaces/core/homeassistant/helpers/entity.py", line 283, in async_write_ha_state
    raise NoEntitySpecifiedError(
homeassistant.exceptions.NoEntitySpecifiedError: No entity id specified for entity Incidents
```

Reply:
Yes something is wrong in that case, maybe the unique id or something? You will have to look at the entity object and compare what data is set in the different sessions

Reply:
Another question: do you really need the try except? Cant you do member validation?

Reply:
Hmm latest code doesn't seem to always update the status, once it fails once (because of no identity id) the sensor keeps being unavailable, not wanted behavior. The former solution was rock solid stable...

Reply:
I think it just hid the issue previously

Reply:
It looked like async_update() was called before self._entity_id of sensor was assigned, which is strange.
Anyway I rewrote it a bit and it seems ok now, is this what you meant by member validation?

Reply:
Yes

I think your issue is that you ask hass to refresh state when you add the entity

    async_add_entities(
         [IncidentsSensor(coordinator, entry)],
         True, <-------
     )

Reply:
Yes, I got the idea it depends how fast it starts if I ran in the issue or not.
Another line to be removed, there is nothing left if we go on like this, I still learn a lot about my own code. :smile:

Reply:
Im improving my own integrations based on what I learn from reviewing others so its a win win :)

Reply:
Do you agree it's ready to merge with dev branch? So I add and rework the switch and binary_sensor code too.

Reply:
I haven't looked at the tests but the code has shaped up well